### PR TITLE
173: Interface to Supervised MLJ Models

### PR DESCRIPTION
**Pull request by RaunoArike**
This pull request adds an interface to support machine learning models from the [MLJ model registry](https://alan-turing-institute.github.io/MLJ.jl/dev/list_of_supported_models/) and documents the possibilities to make them compatible with the counterfactual generators currently implemented in the package. This is an important addition to the package as it provides the basic infrastructure for significantly increasing the scope of the models supported by the package, as well as documentation about the models that can be integrated with the package.

My conclusion based on the research done while working on this PR is that a one-size-fits-all class for all MLJ models is not the right way to handle MLJ models and that taking a more model-specific approach makes more sense. For this reason, this pull request adds support for compatible models from the `EvoTrees.jl` package instead of adding support for all MLJ models at once.

### Changes Made

- Implemented a new abstract type for models from the MLJ library.
- Added a new documentation file in `students_documentation/MLJ_model_report.md` that discusses the feasibility of adding support for various models from the MLJ registry.
- Implemented a prelimenary base type for EvoTrees. The implementation will change during the process of making EvoTrees compatible with our counterfactual generators, which will be tackled as part of another pull request.

### Testing

```
Tested the interface for `EvoTrees.jl` models on all synthetic datasets supported by the package.
```

### CODE REVIEW

File: src/models/differentiable/differentiable.jl [hunk 1]
```diff
@@ -9,6 +9,12 @@ Base type for differentiable models written in pure Julia.
 """
 abstract type AbstractDifferentiableJuliaModel <: AbstractDifferentiableModel end
 
+"""
+Base type for models from the MLJ library.
+"""
+abstract type AbstractMLJModel <: AbstractDifferentiableModel end
+
```
Comment (kmariuszk) [inline src/models/differentiable/differentiable.jl:16]: Is there actually a need to construct this struct? You created it only for the EvoTree model so far (which is differentiable afaik), so why don't just create this as a child class of an `AbstractDifferentiableModel`? I feel like we would need to create another struct for Flux models if we want to divide it that way.

File: src/models/differentiable/evotree_model.jl [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+This type provides a basic interface to gradient-boosted tree models from the MLJ library.
+However, this is not be the final version of the interface: full support for EvoTrees has not been implemented yet and the `logits` and `probs` methods will be changed in the process of doing that if needed.
+"""
+
+"""
+    EvoTreeModel <: AbstractMLJModel
+
+Constructor for gradient-boosted decision trees from the EvoTrees.jl library. 
+"""
+struct EvoTreeModel <: AbstractDifferentiableModel
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:16]: And here you're not even using this `AbstractMLJModel` struct, also the comment is wrong

File: src/models/differentiable/evotree_model.jl [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    model::Any
+    likelihood::Symbol
+    function EvoTreeModel(model, likelihood)
+        if likelihood ∈ [:classification_binary, :classification_multi]
+            new(model, likelihood)
+        else
+            throw(
+                ArgumentError(
+                    "`type` should be in `[:classification_binary, :classification_multi]1.
+                    Support for regressors has not been implemented yet.`"
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:26]: I suppose there's a typo (this "1" at the end of the 25th line)

File: src/models/differentiable/evotree_model.jl [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        end
+    end
+end
+
+"""
+Outer constructor method for EvoTreeModel.
+"""
+function EvoTreeModel(model::Any; likelihood::Symbol=:classification_binary)
+    return EvoTreeModel(model, likelihood)
+end
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:38]: This function looks to be quite random to me, in my opinion, it can cause more confusion from a user perspective rather than benefit them.

File: src/models/differentiable/evotree_model.jl [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+- `M::EvoTreeModel`: The EvoTree model.
+- `X::AbstractArray`: The feature vector for which the predictions are made.
+
+# Returns
+- `p::Matrix`: A matrix of probability scores for each output class for each data point in X.
+
+# Example
+probabilities = Models.probs(M, X) # calculates the probability scores for each output class for each data point in X.
+"""
+function probs(M::EvoTreeModel, X::AbstractArray{<:Number,2})
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:81]: What is this `X::AbstractArray{<:Number,2}` notation?

File: src/models/differentiable/evotree_model.jl [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    output = MLJBase.predict(M.model, DataFrame(X', :auto))
+    p = pdf(output, classes(output))'
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:83]: Are you sure you need to use this `'` operation? It's an adjoint operation, not a transpose (`transpose()` in Julia) (it caused a few errors on my end when working with PyTorch models)

Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:83]: Is `pdf()` a native Julia function? If not, it should be written in the form of `package.pdf()`

File: src/models/differentiable/evotree_model.jl [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    return p
+end
+
+"""
+    probs(M::EvoTreeModel, X::AbstractArray{<:Number, 3})
+
+Works the same way as the probs(M::EvoTreeModel, X::AbstractArray{<:Number, 2}) method above, but handles 3-dimensional rather than 2-dimensional input data.
+"""
+function probs(M::EvoTreeModel, X::AbstractArray{<:Number,3})
+    output = SliceMap.slicemap(x -> probs(M, x), X; dims=[1, 2])
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:105]: I don't fully understand what's going on right here. I guess it could be useful to add a comment before that line and explain more or less what the `SliceMap.slicemap()` operation does and why the `dims` array is `[1, 2]`

File: src/models/differentiable/evotree_model.jl [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    p = pdf(output, classes(output))
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:106]: I suppose this `classes()` function is from the `MLJBase` package, thus please change it to the form of `MLJBase.classes(output)`

File: src/models/differentiable/evotree_model.jl [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+"""
+This type provides a basic interface to gradient-boosted tree models from the MLJ library.
+However, this is not be the final version of the interface: full support for EvoTrees has not been implemented yet and the `logits` and `probs` methods will be changed in the process of doing that if needed.
+"""
+
+"""
+    EvoTreeModel <: AbstractMLJModel
+
+Constructor for gradient-boosted decision trees from the EvoTrees.jl library. 
+"""
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:15]: It would be useful to mention what arguments you need to pass to create this object

File: src/models/differentiable/evotree_model.jl [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    model = EvoTrees.EvoTreeClassifier(kwargs...)
+    mach = machine(model, X, y)
+
+    return EvoTreeModel(mach, data.likelihood)
+end
+
+"""
+    train(M::EvoTreeModel, data::CounterfactualData; kwargs...)
+
+Fits the model `M`` to the data in the `CounterfactualData` object.
```
Comment (kmariuszk) [inline src/models/differentiable/evotree_model.jl:134]: Typo, it should be `M` instead of `M``

File: students_documentation/MLJ_model_report.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+
+### BetaML.jl
+
+| **Model** | **Analysis** |
+| -------- | ------- |
+| DecisionTreeClassifier, RandomForestClassifier | Currently incompatible with the gradient-based generators implemented in our package. However, counterfactual explanations can be generated for these models using the Feature Tweak generator we implemented. The models could also be made compatible with gradient-based generators in the future, possibly through the use of [probability calibration](https://scikit-learn.org/stable/modules/calibration.html). |
+| NeuralNetworkClassifier, LinearPerceptron | Currently incompatible with all generators. However, it should be possible to make these models compatible with all gradient-based generators once the best way to access their gradients has been figured out: all gradient-based generators need gradient-access and currently leverage `Zygote.jl` for that, but `Zygote.jl` does not work for MLJ models, so an alternative solution has to be found. |
+| KernelPerceptron | Currently incompatible with all generators, since the model is not differentiable. The possibility of using probability calibration should be explored. |
```
Comment (kmariuszk) [inline students_documentation/MLJ_model_report.md:20]: Can't we use FeatureTweak for this one?

**Review by kmariuszk — CHANGES_REQUESTED** [review]
Well done! There are some minor issues that I pointed out in the implementation of `EvoTreeModel` but besides that, everything looks good. You made an amazing report, congrats to you!

File: src/models/differentiable/differentiable.jl (hunk 1 above)
Comment (RaunoArike) [reply src/models/differentiable/differentiable.jl:16]: I think it's still worth having this struct, as either us or the client will add more MLJ models in the future. When adding generator support for these models, an abstract class will be needed for them, as the gradient information of MLJ models will be retrieved differently from the way the methods currently retrieve gradient information of the Flux models. However, I can remove this struct for now and it can be added once the package supports more MLJ models, if you think that's the cleaner solution.

File: src/models/differentiable/evotree_model.jl (hunk 4 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:38]: The outer constructor is required because of some Julia quirks, if you look at FluxModel and DeepEnsemble, they also have the same kinds of outer constructors.

File: src/models/differentiable/evotree_model.jl (hunk 3 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:26]: Thanks, fixed.

File: src/models/differentiable/evotree_model.jl (hunk 2 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:16]: Right, good catch. I'll change this based on our final decision on whether to retain the AbstractMLJModel class or not.

File: src/models/differentiable/evotree_model.jl (hunk 5 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:81]: MLJ models require slightly different `probs()` methods depending on whether the input is 1-, 2- or 3-dimensional. The 2 here specifies that the method is called only for 2-dimensional arrays.

File: src/models/differentiable/evotree_model.jl (hunk 6 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:83]: Good catch, it comes from `MLJBase.jl`. Fixed.

File: students_documentation/MLJ_model_report.md (hunk 11 above)
Comment (RaunoArike) [reply students_documentation/MLJ_model_report.md:20]: Assuming you mean KernelPerceptron - I think not, Feature Tweak only works on trees as far as I know.

File: src/models/differentiable/evotree_model.jl (hunk 10 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:134]: Thanks, fixed.

File: src/models/differentiable/evotree_model.jl (hunk 9 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:15]: Sure, will fix.

File: src/models/differentiable/evotree_model.jl (hunk 8 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:106]: Thanks, fixed.

File: src/models/Models.jl [hunk 12]
```diff
@@ -37,6 +37,19 @@ const model_catalogue = Dict(
...
+A dictionary containing both trainable and non-trainable machine learning models.
+"""
+const all_models_catalogue = Dict(
+    :Linear => Linear,
+    :MLP => FluxModel,
+    :DeepEnsemble => FluxEnsemble,
+    :LaplaceRedux => LaplaceReduxModel,
+    :EvoTree => EvoTreeModel,
+)
+
```
[the remark this answers is not in the document]
Comment (kmariuszk) [reply src/models/Models.jl:52]: Yeah, I think it might be necessary to change to avoid ambiguity. Otherwise, you can try to change the name of this variable to something like `trainable_and_non_trainable_models_catalogue` to make it very explicit what's the difference between `model_catalogue` and `all_models_catalogue`.

File: src/models/differentiable/evotree_model.jl (hunk 5 above)
Comment (kmariuszk) [reply src/models/differentiable/evotree_model.jl:81]: Alr, thanks for clarification

File: students_documentation/MLJ_model_report.md (hunk 11 above)
Comment (kmariuszk) [reply students_documentation/MLJ_model_report.md:20]: Yeah, that's what I meant. Okay, thanks for the clarification!

File: src/models/Models.jl (hunk 12 above)
[the remark this answers is not in the document]
Comment (RaunoArike) [reply src/models/Models.jl:52]: Wait, I completely forgot about ctrl + shift + f, I can obviously change it without problems. Will be fixed soon.

File: src/models/differentiable/differentiable.jl (hunk 1 above)
Comment (kmariuszk) [reply src/models/differentiable/differentiable.jl:16]: I don't know much about `LaplaceReduxModel` but maybe something like `CustomDifferentiableModel`? It'd be cool to have a different parent class for each package we're using models from + one additional class for our own implemented models. At the end of the day, we're using those parents' class only for the purpose of building a clear hierarchy, there's no functionality behind it.

File: src/models/differentiable/evotree_model.jl (hunk 6 above)
Comment (kmariuszk) [reply src/models/differentiable/evotree_model.jl]: Good, lmk what are the results

File: src/models/differentiable/evotree_model.jl (hunk 7 above)
Comment (kmariuszk) [reply src/models/differentiable/evotree_model.jl:114]: Yep, it's much cleaner now, thanks.

File: src/models/differentiable/evotree_model.jl (hunk 4 above)
Comment (kmariuszk) [reply src/models/differentiable/evotree_model.jl:45]: Are you sure it's required? I commented it out, ran tests and it worked just fine.

File: src/models/differentiable/differentiable.jl (hunk 1 above)
Comment (RaunoArike) [reply src/models/differentiable/differentiable.jl:16]: I changed it, let me know whether it looks fine to you now.

**Comment by kmariuszk** [comment]
I'm glad you incorporated all suggested changes in your code but you missed updating `.md` files with the new naming of `model_catalogue`. Also, please respond to my comment on `src/models/differentiable/evotree_model.jl` additional constructor function.

File: src/models/differentiable/evotree_model.jl (hunk 4 above)
Comment (RaunoArike) [reply src/models/differentiable/evotree_model.jl:45]: Okay, so it's not required, but it allows us to define an EvoTreeModel without specifying the likelihood - structs can't include default arguments. I don't think we strictly need this, but since the other classes defined by Patrick all have an outer constructor for this reason, maybe we should leave it in?

**Comment by RaunoArike** [comment]
> 

I responded to your constructor comment. Regarding the `.md` files - do you know if there's a way to make the `.qmd` files update the `.md` files automatically? I remember the client saying that he never needs to touch the `.md` files himself.

**Comment by kmariuszk** [comment]
> > 
> 
> I responded to your constructor comment. Regarding the `.md` files - do you know if there's a way to make the `.qmd` files update the `.md` files automatically? I remember the client saying that he never needs to touch the `.md` files himself.

IIRC, they use the Quarto VS code extension which allows you to render `.md` files automatically but I remember it didn't work well on my machine so if you'd run into any issues I'd either suggest changing this manually or reaching our client out and ask them to re-rerender all docs by themselves at the end of the project.

File: src/models/differentiable/evotree_model.jl (hunk 4 above)
Comment (kmariuszk) [reply src/models/differentiable/evotree_model.jl:45]: Yes, that's a good point. Let's leave it in.

**Review by kmariuszk — APPROVED** [review]
LGTM

**Review by VincentPikand — APPROVED** [review]
LGTM
